fastapi_hbr/routing.py

```diff
@@ -0,0 +1,149 @@
+from typing import Any, Callable, Tuple, Type
+
+from fastapi import APIRouter
+from fastapi.routing import APIRoute
+from fastapi.types import DecoratedCallable
+from starlette.datastructures import URL
+from starlette.exceptions import HTTPException
+from starlette.responses import PlainTextResponse
+from starlette.responses import RedirectResponse
+from starlette.routing import Match
+from starlette.types import Receive, Scope, Send
+
+
+async def handle_non_existing_version(scope: Scope, receive: Receive, send: Send):
+    if "app" in scope:
+        raise HTTPException(
+            406,
+            f"Requested version {scope['requested_version']} does not exist. "
+        )
+
+    response = PlainTextResponse("Not Acceptable", status_code=406)
+    await response(scope, receive, send)
+
+
+class HeaderVersionedAPIRoute(APIRoute):
+
+    @property
+    def endpoint_version(self) -> str:
+        return str(self.endpoint.__api_version__)
+
+    def is_version_matching(self, scope: Scope) -> bool:
+        requested_version = scope["requested_version"]
+        return self.endpoint_version == requested_version
+
+    def matches(self, scope: Scope) -> Tuple[Match, Scope]:
+        match, child_scope = super().matches(scope)
+
+        if match == Match.NONE or match == Match.PARTIAL:
+            return match, child_scope
+
+        if self.is_version_matching(scope):
+            return Match.FULL, child_scope
+
+        return Match.PARTIAL, child_scope
+
+    async def handle(self, scope: Scope, receive: Receive, send: Send) -> None:
+        if not self.is_version_matching(scope):
+            await handle_non_existing_version(scope, receive, send)
+            return
+        await super().handle(scope, receive, send)
+
+
+class HeaderVersionedAPIRouter(APIRouter):
+    def __init__(self, *args, route_class: Type[HeaderVersionedAPIRoute] = HeaderVersionedAPIRoute, **kwargs) -> None:
+        self.registered_versions: set[str] = set()
+        super().__init__(*args, route_class=route_class, **kwargs)
+
+    def version(
+            self, api_version: str
+    ) -> Callable[[DecoratedCallable], DecoratedCallable]:
+        self.registered_versions.add(api_version)
+
+        def decorator(func: DecoratedCallable) -> DecoratedCallable:
+            func.__api_version__ = api_version  # type:ignore
+            return func
+
+        return decorator
+
+    def include_router(
+            self,
+            router: "APIRouter",
+            *args: Any,
+            **kwargs: Any
+    ) -> None:
```

Comment:
Considering that this is a public interface, we'll need to expand all args kwargs or use typing tricks (same goes for all other args and kwargs). Otherwise our users won't be able to use this lib comfortably with type checking. Either way let's just add todos and handle it later.
```suggestion
    # TODO: Make sure we show the entire interface of the function to the users later
    def include_router(
            self,
            router: "APIRouter",
            *args: Any,
            **kwargs: Any
    ) -> None:
```

Reply:
I guess copypasting the whole interface of original function works?
And same for other overrided methods (eg couple of __init__s)?

Reply:
Yes. However, there's also the alternative approach which could be preferable: using typing to say that your interface is the same as the interface in the original while having args and kwargs yourself. It makes your code uglier but will evolve better once fastapi adds anything. 

I guess my rule of thumb is: use typing fuckery if arg types and names are the same and use full copying if anything's different.